### PR TITLE
ImageInfo adds a new getter named sizeBytes to decouple ImageCache and ui.Image

### LINKED ISSUE #86402: Extend ImageCache to handle Custom ImageInfo
## Background
We used external texture to display the images in our flutter app. On the Android platform, we draw the `Bitmap` onto the `SurfaceTexture`, and on the iOS platform, we convert the `UIImage` to `CVPixelBufferRef`, and finally use the `Texture` Widget to display the image. We do this mainly based on the following considerations：
1. We hope that any `Bitmap` or `UIImage` can be efficiently displayed to flutter app. For example, we need to customize a photo list page, but the native platform API only provide us `Bitmap` or `UIImage` as photo.
2. Reuse the native platform image loader. the native image loader, such as Glide on the Android platform and SDWebImage on the iOS platform, have implemented a complete set of caching solutions (memory cache, disk cache) and image decoding solutions. In addition, our CDN cropping optimization is also implemented in the native image loader. We hope that flutter app can directly reuse these capabilities.
3. Compared with ui.Image, the memory of external texture is not managed by GC, we can precisely control the timing of memory release

## Use case
We are currently refactoring this solution, we want to let the external texture can benefit from the `ImageCache` logic.
We hope that `ImageCache` can no longer dependent on ui.Image, so that we can customize ImageInfo and ImageProvier to cache external texture images in ImageCache.

## Proposal
let `ImageCache` can no longer dependent on ui.Image，then users can cache custom `ImageInfo`

Related code：
https://github.com/flutter/flutter/blob/31de052e3fddb4909ff88172a6937db7f3dc49db/packages/flutter/lib/src/painting/image_cache.dart#L414

#### Expected
`ImageInfo` add a property named `sizeBytes`, then `ImageCache` use it.

```
 sizeBytes = info.sizeBytes; 
```

#### Actual
```
 sizeBytes = info.image.height * info.image.width * 4; 
```

### PR DESCRIPTION
**Pull request by ColdPaleLight**
We used external texture to display the images in our flutter app. On the Android platform, we draw the Bitmap onto the SurfaceTexture, and on the iOS platform, we convert the UIImage to CVPixelBufferRef, and finally use the Texture Widget to display the image. 
We want to let the external texture can benefit from the ImageCache logic.We hope that ImageCache can no longer dependent on ui.Image, so that we can customize ImageInfo and ImageProvier to cache external texture images in ImageCache.

*List which issues are fixed by this PR. You must list at least one issue.*
Fixes https://github.com/flutter/flutter/issues/86402

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by wongkoo** [comment]
👏🏿 This small change can bring a lot of scalability benefits, awesome

**Comment by ColdPaleLight** [comment]
Hi，@dnfield 
Would you mind reviewing this PR？
Thanks

**Comment by goderbauer** [comment]
Can you please add a test to this PR?

**Review by dnfield — CHANGES_REQUESTED** [review]
This needs a test.

It's also not entirely clear to me what the use case is here. `ImageInfo` must have an `image`, which must have a `width` and `height`. 

It would probably be better to expose `sizeBytes` on `dart:ui.Image` instead.

**Comment by ColdPaleLight** [comment]
## Background
We used external texture to display the images in our flutter app. On the Android platform, we draw the Bitmap onto the SurfaceTexture, and on the iOS platform, we convert the UIImage to CVPixelBufferRef, and finally use the Texture Widget to display the image. We do this mainly based on the following considerations：
1. We hope that any Bitmap or UIImage can be efficiently displayed to flutter app. For example, we need to customize a photo list page, but the native platform API only provide us Bitmap or UIImage as photo.
2. Reuse the native platform image loader. the native image loader, such as Glide on the Android platform and SDWebImage on the iOS platform, have implemented a complete set of caching solutions (memory cache, disk cache) and image decoding solutions. In addition, our CDN cropping optimization is also implemented in the native image loader. We hope that flutter app can directly reuse these capabilities.
3. Compared with ui.Image, the memory of external texture is not managed by GC, we can precisely control the timing of memory release

Use `Texture` Widget to display the image, we only need textureId, width, height, so we have customized a `TextureImageInfo` to represent this image. but  `ImageInfo` must have an `ui.Image`，so we had to give `ImageInfo` a dummy
```
ui.Image _dummy;

FutureOr<TextureImageInfo> createTextureImageInfo(
    {int textureId, int width, int height}) async {
  if (_dummy != null) {
    return TextureImageInfo(
        textureId: textureId, width: width, height: height, image: _dummy);
  }

  _dummy = await _createImage(1, 1);
  return TextureImageInfo(
      textureId: textureId, width: width, height: height, image: _dummy);
}

Future<ui.Image> _createImage(int width, int height) async {
  final Completer<ui.Image> completer = Completer<ui.Image>();
  ui.decodeImageFromPixels(
    Uint8List.fromList(
        List<int>.filled(width * height * 4, 0, growable: false)),
    width,
    height,
    ui.PixelFormat.rgba8888,
        (ui.Image image) {
      completer.complete(image);
    },
  );
  return completer.future;
}

// ignore: must_be_immutable
class TextureImageInfo extends ImageInfo {


  TextureImageInfo({
    this.textureId,
    this.width,
    this.height,
    ui.Image image,
    double scale = 1.0,
    String debugLabel,
  }) : super(
    image: image,
    scale: scale,
    debugLabel: debugLabel,
  );

  int textureId;

  int width;

  int height;

  ......

}

```
Now, we can create `TextureImageInfo` in the following way
```
TextureImageInfo imageInfo = await createTextureImageInfo(textureId:textureId,width:width,height:height)
```

Then we customized an ImageProvider to provide TextureImageInfo，and copy the code of ImageCache and Image Widget and modify their dependence on ui.Image. It works well.

## Expectation
At present, our solution needs to copy the code of `ImageCache` . The reason is that the `image` in our `TextureImageInfo` is dummy, so the correct sizeBytes cannot be calculated. If `ImageInfo` has a `sizeBytes` method, then we can provide the correct sizeBytes without copying the code of `ImageCache` .

**Comment by dnfield** [comment]
Supporting texture based images is an interesting idea. There's also a bug open for supporting other graphics related things in the cache - https://github.com/flutter/flutter/issues/83801

Let me think about this one a little bit.

**Comment by dnfield** [comment]
Can you explain a bit more about how you're getting the size of the texture?

**Comment by ColdPaleLight** [comment]
Hi, @dnfield 

Usually we use the size of Bitmap/CGImage as the size of the texture, but sometimes we can also specify the size of the texture

Android：
```Java
Bitmap bitmap = getBitmapFromSomeWhere();

int textureWidth = bitmap.width();
int textureHeight = bitmap.height();
// we can also specify the size of the texture like this:
// int textureWidth = expectTextureWidthFromSomeWhere;
// int textureHeight = expectTextureHeightFromSomeWhere;

TextureRegistry.SurfaceTextureEntry entry =  textureRegistry.createSurfaceTexture();
Surface surface = new Surface(textureEntry.surfaceTexture());
textureEntry.surfaceTexture().setDefaultBufferSize(width, height);
if (surface != null && surface.isValid()) {
    try {
        Canvas canvas = surface.lockCanvas(null);
        Rect mSrcRect = new Rect(0, 0, bitmap.getWidth(), bitmap.getHeight());
        Rect mDestRect = new Rect(0, 0, width, height);
        canvas.drawBitmap(imageBitmap, mSrcRect, mDestRect, null);
        surface.unlockCanvasAndPost(canvas);
    } catch (Exception e) {
        e.printStackTrace();
        onLoadFailed(imageBitmap, needRecycle);
    }
}

Map<String,Object> result = new HashMap();
result.put("textureId", textureEntry.id());
result.put("width", textureWidth);
result.put("height", textureHeight);

// then send result to flutter
```
The logic of iOS is basically the same as Android，we convert `CGImage` to `CVPixelBuffer`，and size of the texture is the size of the CVPixelBuffer.

```Objective-C
- (CVPixelBufferRef)createPixelBufferFromImage:(CGImageRef)image andSize:(CGSize)size {
    NSDictionary *options = @{
                              (__bridge NSString *)kCVPixelBufferIOSurfacePropertiesKey : @{},
                              (__bridge NSString *)kCVPixelBufferCGImageCompatibilityKey: @NO,
                              (__bridge NSString *)kCVPixelBufferCGBitmapContextCompatibilityKey: @NO
                              };
    
    CVPixelBufferRef pxbuffer = NULL;
    
    
    CGSize imageSize = size;
    if (CGSizeEqualToSize(size, CGSizeZero)) {
        imageSize = CGSizeMake(CGImageGetWidth(image),  CGImageGetHeight(image));
    }
    
    CVReturn status = CVPixelBufferCreate(kCFAllocatorDefault,
                                          imageSize.width,
                                          imageSize.height,
                                          kCVPixelFormatType_32BGRA,
                                          (__bridge CFDictionaryRef) options,
                                          &pxbuffer);
    if (status!=kCVReturnSuccess) {
        NSLog(@"Operation failed");
    }
    NSParameterAssert(status == kCVReturnSuccess && pxbuffer != NULL);
    
    CVPixelBufferLockBaseAddress(pxbuffer, 0);
    void *pxdata = CVPixelBufferGetBaseAddress(pxbuffer);
    
    CGColorSpaceRef rgbColorSpace = CGColorSpaceCreateDeviceRGB();
    CGContextRef context = CGBitmapContextCreate(pxdata, imageSize.width,
                                                 imageSize.height, 8, CVPixelBufferGetBytesPerRow(pxbuffer), rgbColorSpace,
                                                 kCGBitmapByteOrder32Little | kCGImageAlphaPremultipliedFirst);
    
    
    NSParameterAssert(context);
    
    CGContextDrawImage(context, CGRectMake(0, 0, imageSize.width,
                                           imageSize.height), image);
    CGColorSpaceRelease(rgbColorSpace);
    CGContextRelease(context);
    
    CVPixelBufferUnlockBaseAddress(pxbuffer, 0);
    return pxbuffer;
}
```

File: packages/flutter/lib/src/painting/image_stream.dart [hunk 1]
```diff
@@ -98,6 +98,9 @@ class ImageInfo {
   /// the image.
   final ui.Image image;
 
+  /// The size of raw image pixels in bytes
```
Comment (dnfield) [inline packages/flutter/lib/src/painting/image_stream.dart:101]:
```suggestion
  /// The size of raw image pixels in bytes.
```

**Review by dnfield — APPROVED** [review]
LGTM. I can't think of a great reason this API is harmful, and subclassing `ui.Image` could do bad things.

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/86555

**Comment by goderbauer** [comment]
@ColdPaleLight Can you rebase this PR to make the checks pass?

**Comment by ColdPaleLight** [comment]
@googlebot I consent.

**Comment by ColdPaleLight** [comment]
@dnfield @goderbauer All checks have passed. Can you add the label "waiting for tree to go green" again?

**Pull request merged**